### ISSUE TITLE
feat: Lazy load routers

### ISSUE THREAD

**Issue by zomars** [trpc/trpc#4129]
### Describe the feature you'd like to request

Basically `next/dynamic` but for the backend. 

Recently our tRPC router has been growing significantly for many reasons but mostly due to heavy third party SDKs. The problem with this is that even if we're not calling that specific procedure we still load ALL the router dependencies when calling any procedure.

This impacts performance and cold boots significantly.

### Describe the solution you'd like to see

Splitting big tRPC routers into lazy loadable chunks could help loading only what is needed when a procedure is called. I would imagine some pseudo code like this:

![image](https://user-images.githubusercontent.com/3504472/229645485-315a4d62-2044-49ef-b420-0b33bf7fd503.png)

### Describe alternate solutions

Other aternatives involve moving all the procedure code into a separate file a use imports for that [as done in here](https://github.com/calcom/cal.com/pull/7889/commits/6ce440851e172db3f4c7ae720e8396c9201eb316).

Also [lazy loading only third party libraries](https://github.com/calcom/cal.com/pull/7889/commits/a2ff4d9a0cac732af3cd842988feaad471187488) (like Stripe, Google, etc.) seems to help. But it would be a much nicer DX being able to simply lazy load full routers and support it natively.

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

<sub>[TRP-19](https://linear.app/trpc/issue/TRP-19/feat-lazy-load-routers)</sub>

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/4129">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/4129/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/trpc/trpc/issues/4129/pledge.svg">
</picture>
</a>

**Comment by KATT** [comment]
Cool. I'm optimistic we could support this. As a workaround, anyone who approaches this problem should be able to achieve the same perf gains using the SOA-approach: https://github.com/trpc/trpc/tree/main/examples/soa


## An outline of how this could work internally:

**Rough how it currently works:**

- We store a dictionary with all procedures in a flat record, e.g. `{ 'post.byId': () => { /* ... */ `
- When a query is done to `post.byId`, we do a simple lookup to this record at (if I remember correctly) `router._def.procedures['post.byId']`
 
For this to work, our routers need to be a bit smarter and be able to resolve stuff and cache already resolved `import`s. We also need to do split up the incoming procedure's name and by `.` to recursively resolve the path.

**Potential "problematic" code paths**

`createCaller()`  which is currently isn't async and all other server-side paths. We'd need async equivalents that uses a proxy object?

**Comment by Nick-Lucas** [comment]
We had some discussion (no decisions yet) internally on this, so just dumping some of my thoughts now here for wider visibility. **tl;dr** is I'm sceptical this is the right solution, but that doesn't undermine the problem existing and needing a solution.

So firstly a few assertions from other bits of information I've read on this issue, please correct me if I'm wrong:

* The problem is a mixture of overall bundle size, and startup evaluation of the bundle. Combined causing slow cold starts on a serverless lambda
* Currently the API is a single bundle file including dependencies, weighing around 2mb
* The vast majority of the bundle size is going to be 3rd party dependencies. The actual application code is probably quite small

#### So I'm sceptical lazy loading is the true solution for a few reasons:

* If bundle size is a problem, then code-splitting won't fix it, because a lambda would need to download all the "lazy" modules in the image before startup can occur. 2mb isn't that weighty though so I'm guessing not the primary issue here?
* If startup evaluation is a problem, then lazy loading will probably have minimal effect, this because most routers will consume the expensive dependencies such as ORMs and service SDKs, and so you'll end up loading those anyway most of the time.
* Also if startup evaluation is a problem, then bundling less into the bundle would probably be wise, and doesn't necessarily need to be done through lazy loading. I distribute a 20mb lambda at work, but with node_modules part of the distribution rather than bundled, and have experienced no startup delays which impact UX
* If the sheer number and amount of code in service classes from a shared library are the problem, then a DI approach could probably be devised via createContext/middlewares which lazy-loads the service classes on demand. I actually have the opinion that tRPC should offer something in this space, but that's another topic. As things stand createContext is doing a lot of cross-cutting work so really could be part of the problem.

#### The solution to these problems in my view would be to break up the tRPC API into multiple skinnier APIs

...then merge the definitions for the client, and use a special links setup (or API Gateway equivalent) to route appropriately.  @juliusmarminge has already POC'd this and it doesn't look that hard for us to provide some little abstractions in the form of special links, and a way to merge/prefix AppRouter definitions.

This would scale much better to truly massive APIs which are challenging the size limits of a lambda, or complexity limits of Typescript, while also reducing bundle size, and allowing you to chop up the API along dependency boundaries - so one API has prisma, another API doesn't, etc. 

I don't think this is a perfect idea, inevitably dependency boundaries will exist throughout a API's tree rather than neatly by top-level router, and deep merging routers would probably be off the table. 

#### In summary

...there's really nothing too offensive about adding lazy loading, and maybe my above reasoning is wrong. Also it's probably only a handful of lines of code to implement so we not a huge maintenance headache in the grand scheme. The biggest challenge for me here is the lack of data on the problem given I'm looking in from the outside, and the lack of a prototype to compare data against.

**Comment by zomars** [comment]
Just a heads up. We actually went through and lazy loaded ALL tRPC procedures and helped reducing cold start boots from 16seconds to 2-3 seconds. 

So this feature would've help a lot here.

**Comment by zomars** [comment]
We'll be releasing a case study with more in depth numbers soon.

**Comment by KATT** [comment]
I'm actually unsure if this would've actually pushed the needle the same way the refactor to the SOA approach. The culroit is likely not trpc itself but rather each router's dependencies stacking up

**Comment by Nick-Lucas** [comment]
Similar thoughts here, the long term solution to an API getting too large for serverless is to have multiple functions, not to lazy load bits. Lazy loading would just lengthen the runway so to speak. So if we add more 1st class support for solving these problems it will probably be by providing more primitives for merging several tRPC APIs into one client, that way we encourage and support the best practice solution.

Happy to be proven wrong of course, but I've been following this piece via Julius and gained the understanding that there was essentially a subset of routers with a heavy dependency around (ie.) i18n and carving that off made a big difference.

**Comment by MJomaa** [comment]
From a DX perspective the lazy loading syntax in the initial post is fantastically easy to understand and use.

**Comment by MathiasGruber** [comment]
Agreed, lazy loading would be an awesome feature to have

**Comment by MJomaa** [comment]
Feedback from me.. whe removed tRPC from some endpoints because we saw  huge cold  cold start improvements. (from seconds to ms)

Currently migrating 180 endpoints away. Furthermore it turned out that we needed that hooks layer for some time zone conversions. Still glad we chose tRPC to prototype early on. You will miss the easy prototype when you don't have it anymore.

**Comment by FelipeJz** [comment]
+1 Huge projects will benefit a lot!

**Comment by amritk** [comment]
> #### The solution to these problems in my view would be to break up the tRPC API into multiple skinnier APIs
> ...then merge the definitions for the client, and use a special links setup (or API Gateway equivalent) to route appropriately.

@Nick-Lucas this is exactly what I'm looking for and would be willing to help implement if needed. Is there a separate ticket tracking this feature?

**Comment by maiconcarraro** [comment]
I was facing the same issue w/ cold boots and especially slow transition between pages, and had no idea what was causing this until I used [Trace viewer](https://www.youtube.com/watch?v=PGO2szAye7A) from [@timneutkens](https://github.com/timneutkens) and found it was related to trpc + some procedures importing 3rd parties like opensearch+aws... as a workaround I'm fixing the heavy procedures, but sharing here some info in case someone wants to use the same **Trace viewer** to find out the heavy routers and procedures.

![image](https://github.com/user-attachments/assets/7062f6ac-6995-4d96-b309-08f475933ae1)

**Comment by maiconcarraro** [comment]
@KATT I'm late for your PR https://github.com/trpc/trpc/pull/5489 which added the lazy, in case you want some help to test it I can try on my project using the trace viewer above to compare results, I'm currently using `11.0.0-rc.485` in case it helps.

**Comment by acidoxee** [comment]
Might this be reignited? We've previously migrated off of tRPC because of high cold starts in lambdas, but this might turn back the tide in its favor if we can mitigate those nasty timings.

**Comment by Rototu** [comment]
Besides lambda cold starts, as we're not using trpc in our lambdas, our team is facing issues with starting our Next.js dev servers, as well as with hot reloading.

The fact that the whole router has to be loaded for any page, even for ones that use just one or two lightweight procedures has slowed down our dev server a lot. I am more than willing to help out with testing your PR as well @KATT, please reach out to me if you need any help testing it. 

---

For more context:

Atm, we have this old app that still had all of it's api endpoints using Next's pages router, so I was working on migrating it to using app router + trpc to be in line with our other projects. 

On the same page (first page opened after dev server startup), where some requests were using the new TRPC procedures, vs the old ones using the pages router endpoints, the data fetched using the old endpoint loaded instantly, while the data fetched using TRPC had a few seconds of delay due to the dev server having to build the whole router. 

On our newest app that is fully on TRPC, simply removing one single very bulky procedure with from the router saw our dev server first page loads go down by 7-10 seconds. 

Right now we're addressing this by moving these bulky procedures to our actual BE, a django based api server, but ofc that is not always ideal for us.

**Comment by maiconcarraro** [comment]
@KATT can you confirm which release version to try it? And really appreciate the hard work 🙏

**Comment by KATT** [comment]
`next` tag

**Comment by KATT** [comment]
Let us know how it works!

**Comment by maiconcarraro** [comment]
@KATT tried using `11.0.0-rc.748` and it gave duplicate key while navigating, it does work after some reloads, maybe concurrency?

![Image](https://github.com/user-attachments/assets/f4e179db-a3e5-4b7a-941f-0245b7f60922)

`createTRPCRoutert` is `t.router`, using t3 stack

```
 ○ Compiling /u/[username] ...
 ✓ Compiled /u/[username] in 8.9s
💤 lazy loaded profileRouter
 ⨯ Error: Duplicate key: profile.me
    at async Page...
```

**Comment by KATT** [comment]
Could you make a repro and post an issue?

Should be easy to copy the folder in https://github.com/trpc/trpc/tree/next/examples/lazy-load

**Comment by maiconcarraro** [comment]
@KATT the bug seems to happen only in nextjs, so I used a different template, added + information here https://github.com/trpc/trpc/issues/6469
